src/native/dnmd/src/interfaces/pal.cpp

```diff
@@ -0,0 +1,327 @@
+{
+    return ConvertUtf16ToUtf8(c, buffer, bufferLength, &bufferLength);
+}
+
+template<>
+HRESULT pal::StringConvert<char, WCHAR>::ConvertWorker(char const* c, WCHAR* buffer, uint32_t& bufferLength)
+{
+    return ConvertUtf8ToUtf16(c, buffer, bufferLength, &bufferLength);
+}
+
+#if !defined(__STDC_LIB_EXT1__) && !defined(BUILD_WINDOWS)
+int strcat_s(char* dest, rsize_t destsz, char const* src)
+{
+    assert(dest != nullptr && src != nullptr);
+    (void)::strcat(dest, src);
+    return 0;
+}
+#endif // !defined(__STDC_LIB_EXT1__) && !defined(BUILD_WINDOWS)
+
+// SHA1 implementation
+#if defined(BUILD_WINDOWS)
+namespace
+{
+    struct BCRYPT_ALG_HANDLE_deleter final
+    {
+        void operator()(BCRYPT_ALG_HANDLE h) const noexcept
+        {
+            ::BCryptCloseAlgorithmProvider(h, 0);
+        }
+    };
+
+    using bcrypt_alg_handle = std::unique_ptr<void, BCRYPT_ALG_HANDLE_deleter>;
+
+    struct BCRYPT_HASH_HANDLE_deleter final
+    {
+        void operator()(BCRYPT_HASH_HANDLE h) const noexcept
+        {
+            ::BCryptDestroyHash(h);
+        }
+    };
+
+    using bcrypt_hash_handle = std::unique_ptr<void, BCRYPT_HASH_HANDLE_deleter>;
+}
+
+bool pal::ComputeSha1Hash(span<const uint8_t> data, std::array<uint8_t, SHA1_HASH_SIZE>& hashDestination)
+{
+    BCRYPT_ALG_HANDLE hAlg;
+    if (!BCRYPT_SUCCESS(::BCryptOpenAlgorithmProvider(&hAlg, BCRYPT_SHA1_ALGORITHM, nullptr, 0)))
+    {
+        return false;
+    }
+    bcrypt_alg_handle algHandle { hAlg };
+
+    BCRYPT_HASH_HANDLE hHash;
+    if (!BCRYPT_SUCCESS(::BCryptCreateHash(hAlg, &hHash, nullptr, 0, nullptr, 0, 0)))
+    {
+        return false;
+    }
+    bcrypt_hash_handle hashHandle { hHash };
+    if (!BCRYPT_SUCCESS(::BCryptHashData(hHash, (PUCHAR)(uint8_t const*)data, (ULONG)data.size(), 0)))
+    {
+        return false;
+    }
+
+    return BCRYPT_SUCCESS(::BCryptFinishHash(hHash, hashDestination.data(), (ULONG)hashDestination.size(), 0));
+}
+#elif defined(BUILD_MACOS)
+#include <CommonCrypto/CommonDigest.h>
+
+bool pal::ComputeSha1Hash(span<const uint8_t> data, std::array<uint8_t, SHA1_HASH_SIZE>& hashDestination)
+{
+    static_assert(CC_SHA1_DIGEST_LENGTH == SHA1_HASH_SIZE, "SHA1 hash size mismatch");
+    CC_SHA1(data, data.size(), hashDestination.data());
+    return true;
+}
+
+#elif defined(BUILD_UNIX) && !defined(FEATURE_DISTRO_AGNOSTIC_SSL)
+#include <openssl/sha.h>
+bool pal::ComputeSha1Hash(span<const uint8_t> data, std::array<uint8_t, SHA1_HASH_SIZE>& hashDestination)
+{
+    static_assert(SHA_DIGEST_LENGTH == SHA1_HASH_SIZE, "SHA1 hash size mismatch");
+    SHA1(data, data.size(), hashDestination.data());
+    return true;
+}
+#elif defined(BUILD_UNIX) && defined(FEATURE_DISTRO_AGNOSTIC_SSL)
+namespace
+{
+    struct sha1_func
+    {
+        void (*func)(const uint8_t*, size_t, uint8_t*);
+        sha1_func()
+        {
+            void* img_handle = dlopen("libcrypto.so", RTLD_LAZY);
```

Comment:
I do not think we want to introduce a core runtime dependency on OpenSSL. OpenSSL (and other real crypto stacks) can be locked down in various ways. We do not want to the core runtime to fail to start if they decide to lock down the legacy hash alg. It creates a whole new set of failure modes.

The core runtime has an exception for non-crypto use of the legacy crypto hash algorithms. If we wanted to do something about it, we would take a breaking change to switch to non-crypto hash algorithms.

Reply:
I'll move the SHA1 implementation from CoreCLR into the minipal and use it from here (for the exact same case that it's used in CoreCLR today).